src/v/archival/ntp_archiver_service.cc

```diff
@@ -2026,15 +2026,29 @@ ss::future<ntp_archiver::batch_result> ntp_archiver::wait_all_scheduled_uploads(
       .compacted_upload_result = compacted_result};
 }
 
+model::offset ntp_archiver::max_uploadable_offset() const {
+    // We impose an additional (LSO) constraint on the uploadable offset to
+    // as we need to have a complete index of aborted transactions if any
+    // before we can upload a segment.
+    return std::min(
+      model::prev_offset(_parent.last_stable_offset()),
+      _parent.committed_offset());
+}
+
 ss::future<ntp_archiver::batch_result> ntp_archiver::upload_next_candidates(
   std::optional<model::offset> lso_override) {
-    vlog(_rtclog.debug, "Uploading next candidates called for {}", _ntp);
-    auto last_stable_offset = lso_override ? *lso_override
-                                           : _parent.last_stable_offset();
+    auto max_stable_offset = lso_override
+                               ? *lso_override
+                               : model::next_offset(max_uploadable_offset());
```

Comment:
It's a bit confusing that we're using `prev_offset` inside the `max_uploadable_offset` and then we're using `next_offset`.  Maybe the variable `max_stable_offset` should be called differently(first_unstable_offset or end_eclusive)?

Reply:
I wanted to keep the `max_uploadable_offset()` method name short and clear. `max_uploadable_offset_inclusive` sounds is too long. `first_unstable_offset` doesn't carry enough meaning.

Then, I could still have avoided the `model::next_offset` but then I needed to change all the call sites that specify an lso_override and downstream sites that use it. Error prone too. But probably we should just do that. And, rename lso_override to max_uploadable_offset_override too.

wdyt?

Reply:
Yea it's unfortunate the naming is tied with "stable offset" which itself is confusingly named and tied with Kafka nomenclature.

+1 to renaming the `max_stable_offset` variable to some `end_exclusive`. I think the `max_uploadable_offset()` method name is good though (succinct and clear).